packages/react-server/src/ReactFlightServer.js

```diff
@@ -1237,6 +1237,9 @@ function renderModel(
           task.implicitSlot,
           request.abortableTasks,
         );
+        // The suspended element was outlined, so we're using the same ID for the
+        // original value, thus ensuring that it's deduplicated if it's referenced again.
+        request.writtenObjects.set((value: any), newTask.id);
```

Comment:
There's a couple of things to consider here.

- This value might not be one of the kinds of objects that we stash in this map or even an object at all necessarily. So that would need to be confirmed.
- The thing being executed in the task is really the `model`, not `value`. The value could be something we've already rendered past. It would be safer to do this with just `model`.
- It's possible for the same element to be rendered inside two different "contexts". We removed Server Context but `keyPath` and `implicitSlot` is still a contextual thing. You could render the same element in two different key paths.

E.g. this would be safer:

```
if (wasReactNode && task.keyPath === null && !task.implicitSlot) {
  request.writtenObjects.set(model, newTask.id);
}
```

Reply:
See https://github.com/facebook/react/blob/main/packages/react-server/src/ReactFlightServer.js#L879-L880

Reply:
While investigating whether I could avoid the cast to `any` for `value`, I also thought I could maybe just use `model` instead (which, I later noticed, I would also need to cast). But it turns out that these two are not the same because of `createLazyWrapperAroundWakeable`, and then the deduplication does not work:

```js
{
  value: {
    '$$typeof': Symbol(react.element),
    type: [AsyncFunction: Bar],
    key: null,
    ref: null,
    props: { text: 'bar' }
  },
  model: {
    '$$typeof': Symbol(react.lazy),
    _payload: Promise { 'BAR', status: 'pending' },
    _init: [Function: readThenable]
  }
}
```

Reply:
So maybe this is good enough as a heuristic then?

```js
if (wasReactNode && task.keyPath === null && !task.implicitSlot) {
  request.writtenObjects.set(value, newTask.id);
}
```

I'm not sure whether `wasReactNode` always covers `value` transitively. If not, we could do the same check for `value`, I guess?

Reply:
>It's possible for the same element to be rendered inside two different "contexts". We removed Server Context but keyPath and implicitSlot is still a contextual thing. You could render the same element in two different key paths.

Can you expand on this a bit? Using the unit test as an example, `implicitSlot` is indeed set to `true` [here](https://github.com/facebook/react/blob/fd0da3eef1e23b7dd5071fef21de414da8e5038e/packages/react-server/src/ReactFlightServer.js#L655) for `<Bar />` because it does not have a key path. Which means, using the proposed condition would only work if we move the code a couple lines further down, after the context has been restored. Would that still be correct?